src/afl-fuzz-extras.c

```diff
@@ -370,16 +387,74 @@ static inline u8 memcmp_nocase(u8 *m1, u8 *m2, u32 len) {
 
 }
 
-/* Adds a new extra / dict entry. Used for LTO autodict. */
+/* Removes duplicates from the loaded extras. This can happen if multiple files
+   are loaded */
+
+void dedup_extras(afl_state_t *afl) {
+
+  if (afl->extras_cnt < 2) return;
+
+  u32 i, j, orig_cnt = afl->extras_cnt;
+
+  for (i = 0; i < afl->extras_cnt - 1; i++) {
+
+    for (j = i + 1; j < afl->extras_cnt; j++) {
+
+    restart_dedup:
+
+      // if the goto was used we could be at the end of the list
+      if (j >= afl->extras_cnt || afl->extras[i].len != afl->extras[j].len)
+        break;
+
+      if (memcmp(afl->extras[i].data, afl->extras[j].data,
+                 afl->extras[i].len) == 0) {
+
+        ck_free(afl->extras[j].data);
+        if (j + 1 < afl->extras_cnt)  // not at the end of the list?
+          memmove((char *)&afl->extras[j], (char *)&afl->extras[j + 1],
+                  (afl->extras_cnt - j - 1) * sizeof(struct extra_data));
+        afl->extras_cnt--;
+        goto restart_dedup;  // restart if several duplicates are in a row
+
+      }
+
+    }
+
+  }
+
+  if (afl->extras_cnt != orig_cnt)
+    afl->extras = afl_realloc((void **)&afl->extras,
```

Comment:
Note that `afl_realloc` will never shrink the buffer size. This call has no effect.
(The idea behind `afl_realloc` is, that we reuse buffers as much as possible only reallocate if we really need to -> if the new buffer is larger than what we saw to this day).
Even if extras could grow in this function, the `!=` check would be redundant as it's done in realloc, too.

Reply:
uh true. so what is the correct way?

Reply:
I would just not resize it here, we may add extras later and need to re-resize, right? Alternatively, switch to normal malloc. I can also add an afl_realloc_shrinking if you want?

Reply:
or afl_realloc_exact ... yes I think that would be a good thing to have

Reply:
-> a2739ef ?

Reply:
plus ...  b33306c